pkg/observer/controller.go

```diff
@@ -273,31 +273,76 @@ func (c *controller) dropInProgressNodeGroups(nodeGroups v1.NodeGroupList, cnrs
+        nodeNames := make([]string, 0, len(nodeGroup.List))
+        for _, node := range nodeGroup.List {
+            nodeNames = append(nodeNames, node.Name)
+        }
+        // generate cnr with prefix and use generate name method
+        cnr := generation.GenerateCNR(*nodeGroup.NodeGroup, nodeNames, c.CNRPrefix, c.Namespace)
+        generation.UseGenerateNameCNR(&cnr)
+        generation.GiveReason(&cnr, nodeGroup.Reason)
+        generation.SetAPIVersion(&cnr, apiVersion)
+
+        name := generation.GetName(cnr.ObjectMeta)
+
+        if err := generation.ApplyCNR(c.client, c.DryMode, cnr); err != nil {
+            klog.Errorf("failed to apply cnr %q for nodegroup %q: %s", name, nodeGroup.NodeGroup.Name, err)
+        } else {
+            var drymodeStr string
+            if c.DryMode {
+                drymodeStr = "[drymode] "
+            }
+            klog.V(2).Infof("%ssuccessfully applied cnr %q for nodegroup %q", drymodeStr, name, nodeGroup.NodeGroup.Name)
+            c.CNRsCreated.WithLabelValues(nodeGroup.NodeGroup.Name).Inc()
+        }
+    }
+}
 
-		name := generation.GetName(cnr.ObjectMeta)
+// selectLowestPriorityNodeGroups returns only the node groups at the lowest priority value
+func (c *controller) selectLowestPriorityNodeGroups(changedNodeGroups []*ListedNodeGroups) []*ListedNodeGroups {
+    if len(changedNodeGroups) == 0 {
+        return nil
+    }
+    getPriority := func(ng *v1.NodeGroup) int32 { return max(ng.Spec.Priority, 0) }
+    minPriority := getPriority(changedNodeGroups[0].NodeGroup)
+    for i := 1; i < len(changedNodeGroups); i++ {
+        p := getPriority(changedNodeGroups[i].NodeGroup)
+        if p < minPriority {
+            minPriority = p
+        }
+    }
+    filtered := make([]*ListedNodeGroups, 0, len(changedNodeGroups))
+    for _, changeNodeGroup := range changedNodeGroups {
+        if getPriority(changeNodeGroup.NodeGroup) == minPriority {
+            filtered = append(filtered, changeNodeGroup)
+        }
+    }
+    return filtered
+}
 
-		if err := generation.ApplyCNR(c.client, c.DryMode, cnr); err != nil {
-			klog.Errorf("failed to apply cnr %q for nodegroup %q: %s", name, nodeGroup.NodeGroup.Name, err)
-		} else {
-			var drymodeStr string
-			if c.DryMode {
-				drymodeStr = "[drymode] "
-			}
-			klog.V(2).Infof("%ssuccessfully applied cnr %q for nodegroup %q", drymodeStr, name, nodeGroup.NodeGroup.Name)
-			c.CNRsCreated.WithLabelValues(nodeGroup.NodeGroup.Name).Inc()
-		}
-	}
+// hasLowerPriorityCNRsInProgress returns true if any in-progress CNR belongs to a NodeGroup with
+// a priority lower than the provided minPriority (i.e., must finish before creating higher priorities)
+func (c *controller) hasLowerPriorityCNRsInProgress(minPriority int32, inProg v1.CycleNodeRequestList) bool {
+    if len(inProg.Items) == 0 {
+        return false
+    }
+    // Build a list of valid nodegroups to map CNRs to priorities
+    allNG := c.validNodeGroups()
```

Comment:
Follow-up from @mwhittington21 comment regarding `inProg`, that should apply to all variables here. So this should become `allNodegroups`.